pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -269,15 +276,26 @@ func (e *EndpointController) processNextWorkItem() bool {
 	defer e.queue.Done(eKey)
 
 	err := e.syncService(eKey.(string))
+	e.handleErr(err, eKey)
+
+	return true
+}
+
+func (e *EndpointController) handleErr(err error, key interface{}) {
 	if err == nil {
-		e.queue.Forget(eKey)
-		return true
+		e.queue.Forget(key)
+		return
 	}
 
-	utilruntime.HandleError(fmt.Errorf("Sync %v failed with %v", eKey, err))
-	e.queue.AddRateLimited(eKey)
+	if e.queue.NumRequeues(key) < maxRetries {
+		glog.V(2).Infof("Error syncing endpoints for service %v: %v", key, err)
+		e.queue.AddRateLimited(key)
+		return
+	}
 
-	return true
+	utilruntime.HandleError(err)
```

Comment:
- [ ] Consider moving this statement down two lines until after you've called `workqueue.RateLimitingInterface.Forget`.  
  That way, in case someone installed a handler with `runtime.ErrorHandlers` that panics, you will ensure that the rate limiter is in the proper state for this key.